File: 1631-path-with-minimum-effort/1631-path-with-minimum-effort.cpp
typedef pair<int, pair<int, int>> pipii;

class Solution {
public:
    int minimumEffortPath(vector<vector<int>>& heights) {
        int n = heights.size();
        int m = heights[0].size();

        int deltaRow[] = {-1, 0, +1, 0};
        int deltaCol[] = {0, +1, 0, -1};

        vector<vector<int>> effort(n, vector<int>(m, INT_MAX));
        effort[0][0] = 0;

        // {effort, {row, col}}
        priority_queue<pipii, vector<pipii>, greater<pipii>> pq;
        pq.push({0, {0,0}});

        while(!pq.empty()){
            int eff = pq.top().first;
            int row = pq.top().second.first;
            int col = pq.top().second.second;
            pq.pop();

            if(row == n-1 and col == m-1) return eff;

            for(int i = 0; i < 4; i++){
                int newRow = row + deltaRow[i];
                int newCol = col + deltaCol[i];

                if(newRow >= 0 and newRow < n and newCol >= 0 and newCol < m){
                    int newEffort = max(abs(heights[row][col] - heights[newRow][newCol]), eff);

                    if(newEffort < effort[newRow][newCol]){
                        effort[newRow][newCol] = newEffort;
                        pq.push({newEffort, {newRow, newCol}});
                    }
                }
            }
        }

        return -1;
    }
};

File: 1631-path-with-minimum-effort/1631-path-with-minimum-effort.java
class Solution {
    public int minimumEffortPath(int[][] heights) {
        int n = heights.length;
        int m = heights[0].length;

        int[][] directions = {{1, 0}, {0, 1}, {-1, 0}, {0, -1}};

        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> a[0] - b[0]);
        pq.add(new int[]{0, 0, 0}); // {absDiffrence, row, col}

        int[][] effort = new int[n][m];
        for(int i = 0; i < n; i++){
            Arrays.fill(effort[i], Integer.MAX_VALUE);
        }
        effort[0][0] = 0;


        int minimumEffort = 0;

        while(!pq.isEmpty()){
            int[] top = pq.poll();
            int currentEffort = top[0];
            int r = top[1];
            int c = top[2];

            if(r == n-1 && c == m-1) return currentEffort;

            for(int[] direction : directions){
                int nr = r + direction[0];
                int nc = c + direction[1];

                if(nr >= 0 && nr < n && nc >= 0 && nc < m){
                    int newEffort = Math.max(currentEffort, Math.abs(heights[r][c] - heights[nr][nc]));

                    if(newEffort < effort[nr][nc]){
                        effort[nr][nc] = newEffort;
                        pq.add(new int[]{newEffort, nr, nc});
                    }
                }
            }
        }

        return -1;

    }
}